test_unit.py

```diff
@@ -3600,6 +3613,537 @@ def test_local_state(self):
+        for i in range(255):
+            # Pass in an int type to encode
+            actual = ByteType().encode(i)
+            expected = i.to_bytes(1, byteorder="big")
+            self.assertEqual(actual, expected)
+
+            # Try passing in a bytes type value to encode
+            actual = ByteType().encode(expected)
+            self.assertEqual(actual, expected)
+
+            # Test decoding
+            actual = ByteType().decode(actual)
+            expected = i
+            self.assertEqual(actual, expected)
+
+        # Try to encode a bad byte
+        with self.assertRaises(error.ABIEncodingError) as e:
+            ByteType().encode(256)
+        with self.assertRaises(error.ABIEncodingError) as e:
+            ByteType().encode(-1)
+        with self.assertRaises(error.ABIEncodingError) as e:
+            ByteType().encode((256).to_bytes(2, byteorder="big"))
+        with self.assertRaises(error.ABIEncodingError) as e:
+            ByteType().decode(bytes.fromhex("8000"))
+        with self.assertRaises(error.ABIEncodingError) as e:
+            ByteType().decode((256).to_bytes(2, byteorder="big"))
+
+    def test_address_encoding(self):
+        for _ in range(100):
+            # Generate 100 random addresses as strings and as 32-byte public keys
+            random_addr_str = account.generate_account()[1]
+            addr_type = AddressType()
+            actual = addr_type.encode(random_addr_str)
+
+            expected = encoding.decode_address(random_addr_str)
+            self.assertEqual(actual, expected)
+
+            actual = addr_type.encode(expected)
+            self.assertEqual(actual, expected)
+
+            # Test decoding
+            actual = addr_type.decode(actual)
+            self.assertEqual(actual, expected)
+
+    def test_string_encoding(self):
+        # Test *some* valid combinations of UTF-8 characters
+        chars = string.ascii_letters + string.digits + string.punctuation
+        for _ in range(1000):
+            test_case = "".join(
+                random.choice(chars) for i in range(random.randint(0, 1000))
+            )
+            str_type = StringType()
+            str_len = len(test_case).to_bytes(2, byteorder="big")
+            expected = str_len + bytes(test_case, "utf-8")
+            actual = str_type.encode(test_case)
+            self.assertEqual(actual, expected)
+
+            # Test decoding
+            actual = str_type.decode(actual)
+            self.assertEqual(actual, test_case)
+
+        with self.assertRaises(error.ABIEncodingError) as e:
+            StringType().decode((0).to_bytes(1, byteorder="big"))
+        with self.assertRaises(error.ABIEncodingError) as e:
+            StringType().decode((1).to_bytes(2, byteorder="big"))
+
+    def test_array_static_encoding(self):
+        test_cases = [
+            (
+                ArrayStaticType(BoolType(), 3),
+                [True, True, False],
+                bytes.fromhex("C0"),
```

Comment:
nit: a format like `bytes([0b11000000])` would be easier to understand for these boolean test cases